src/main/java/io/nats/client/impl/NatsConnectionWriter.java

```diff
@@ -119,13 +127,21 @@ boolean isRunning() {
         return running.get();
     }
 
+    private static final NatsMessage END_RECONNECT = new NatsMessage("_end", null, EMPTY_BODY);
+
+    AtomicReference<Mode> lastMode = new AtomicReference<>(Mode.EndReconnect);
+
     void sendMessageBatch(NatsMessage msg, DataPort dataPort, StatisticsCollector stats) throws IOException {
         writerLock.lock();
         try {
             int sendPosition = 0;
             int sbl = sendBufferLength.get();
 
             while (msg != null) {
+                if (msg == END_RECONNECT) {
+                    mode.set(Mode.Outgoing);
+                    return;

```

Comment:
```suggestion
                    msg = msg.next;
                    continue;
```

We can lose all messages in this NatsMessage linked list when on of them is END_RECONNECT.
I created a test in https://github.com/nats-io/nats.java/pull/1445 to demonstrate it